core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/entity/user/User.java

```diff
@@ -54,8 +55,14 @@
     void setSecurityQuestion(SecurityQuestion securityQuestion);
 
     String getSecurityAnswer();
+    
+    String getClearSecurityAnswer();
 
-    void setSecurityAnswer(String securityAnswer);
+    void setEncodedSecurityAnswer(String securityAnswer);
+    
+    void setSecurityAnswer(String securityAnswer,  CipherAlgorithm cipherAlgoritm);
```

Comment:
This method would allow to set  a different `cipherAlgorithm` for the user than the one set for password.
The risk is that `user#getCipherAlgorith` can return the wrong value when used to check password or security question values via `Encryptor#verify`.

I think we should change `User` interface to:

1. remove `cipherAlgorithm` param from `setPassword()` and `setSecurityAnswer()` methods
1. add a separate `setCipherAlgorithm()` method, failing if `cipherAlgorithm` is already set, unless the provided parameter is `null`
1. have `setPassword()` and `setSecurityAnswer()` to internally set `cipherAlgorithm` to the value from conf

Reply:
Yes, totally agree. In fact I didn't add `setCipherAlgorithm()` from `setSecurityAnswer()`implementation.
Going to perform your proposed changes

Reply:
do we also remove from `setEncodedPassword()`? Seems the same case

Reply:
`setEncodedPassword` is used internally by DAO, it is required

Reply:
I've notice that disallowing setting cipher algorithm once set could impact cases like the one in this test
https://github.com/apache/syncope/blob/2_1_X/core/provisioning-java/src/test/java/org/apache/syncope/core/provisioning/java/MappingManagerImplTest.java#L61-L132

Could it be too much restrictive? That is not so usual situation, but could lead to errors if for some reason an admin decides to change password hashing algorithm, with all its consequences...

Reply:
Tests can't be a reason to change code / design.
Just user `ReflectionTestUtils#setField` if nothing else works.

Reply:
Ok, yes, I'll use that. My wonder was to cases when we want to change password cipher algo for some reason. In my experience this happened only once.